assets/server/realmadmin/_form_sms.html

```diff
@@ -375,13 +370,59 @@
         return;
       }
 
+      let errors = new Array();
       let val = ($(target).val() || '').trim();
-      val = val.replace(/\[region\]/g, region);
+
+      if (val.length > {{$realm.SMSTemplateMaxLength}}) {
+        errors.push('SMS Templates must be <= {{$realm.SMSTemplateMaxLength}} characters, currently ' + val.length + ' characters.');
+      }
+
+      // Provide live feedback on errors in the SMS Template construction.
+      if (enxEnabled) {
+        if (!val.includes("[enslink]")) {
+          errors.push("SMS template must contains [enslink]");
+        }
+        if (val.includes("[region]")) {
+          errors.push("cannot contain `[region]`, this is automatically included in [enslink]");
+        }
+        if (val.includes("[longcode]")) {
+          errors.push("cannot contain `[longcode]`, this is automatically included in [enslink]");
+        }
+        val = val.replace(/\[enslink\]/g, ensLink);      
+      } else {
+        if (val.includes("[enslink]")) {
+          errors.push("`[enslink]` is not allowed when Exposure Notifications Express is not enabled.");
+        }
+
+        hasSC = val.includes("[code]");
+        hasLC = val.includes("[longcode]");
+        if (!(hasSC || hasLC) || (hasSC && hasLC)) {
+          errors.push('must contain exactly one of `[code]` or `[longcode]`');
+        }
+
+        val = val.replace(/\[region\]/g, region);
+        val = val.replace(/\[longcode\]/g, longCode);
+      }    
       val = val.replace(/\[code\]/g, shortCode);
-      val = val.replace(/\[expires\]/g, shortExpires);
-      val = val.replace(/\[longcode\]/g, longCode);
+      val = val.replace(/\[expires\]/g, shortExpires);      
       val = val.replace(/\[longexpires\]/g, longExpires);
-      val = val.replace(/\[enslink\]/g, ensLink);
+
+      if (val.length > {{$realm.SMSTemplateExpansionMax}}) {
+        errors.push('Fully expanded SMS Templates must be <= {{$realm.SMSTemplateExpansionMax}} characters, currently ' + val.length + ' characters.');
+      }
+
+      if (errors.length > 0) {
```

Comment:
Why not just call `flash.error` instead of building a custom error list?

Reply:
I actually like this better - it updates live (character counts) and follows the selected message in the editor